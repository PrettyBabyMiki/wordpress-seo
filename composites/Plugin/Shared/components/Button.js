import React from "react";
import styled from "styled-components";
import flow from "lodash/flow";
import omit from "lodash/omit";
import PropTypes from "prop-types";

import colors from "../../../../style-guide/colors.json";
import { Icon } from "./Icon";
import { rgba } from "../../../../style-guide/helpers";

const settings = {
	minHeight: 32,
	verticalPadding: 4,
	borderWidth: 1,
};

const ieMinHeight = settings.minHeight - ( settings.verticalPadding * 2 ) - ( settings.borderWidth * 2 );

/**
 * Returns a component with applied base button styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied base button styles.
 */
export function addBaseStyle( component ) {
	return styled( component )`
		display: inline-flex;
		align-items: center;
		justify-content: center;
		vertical-align: middle;
		border-width: ${ `${ settings.borderWidth }px` };
		border-style: solid;
		margin: 0;
		padding: ${ `${ settings.verticalPadding }px` } 10px;
		border-radius: 3px;
		cursor: pointer;
		box-sizing: border-box;
		font-size: inherit;
		font-family: inherit;
		font-weight: inherit;
		text-align: left;
		overflow: visible;
		min-height: ${ `${ settings.minHeight }px` };

		svg {
			// Safari 10
			align-self: center;
		}

		// Only needed for IE 10+.
		@media all and ( -ms-high-contrast: none ), ( -ms-high-contrast: active ) {
			::after {
				display: inline-block;
				content: "";
				min-height: ${ `${ ieMinHeight }px` };
			}
		}
	`;
}

/**
 * Returns a component with applied focus styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied focus styles.
 */
export function addFocusStyle( component ) {
	return styled( component )`
		&::-moz-focus-inner {
			border-width: 0;
		}

		&:focus {
			outline: none;
			border-color: ${ colors.$color_blue };
			background-color: ${ colors.$color_white };
			box-shadow: 0 0 3px ${ rgba( colors.$color_blue_dark, .8 ) };
		}
	`;
}

/**
 * Returns a component with applied hover styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied hover styles.
 */
export function addHoverStyle( component ) {
	return styled( component )`
		&:hover {
			background: ${ colors.$color_button_hover };
			border-color: ${ colors.$color_button_border_hover };
			color: ${ colors.$color_button_text_hover };
		}
	`;
}

/**
 * Returns a component with applied active styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied active styles.
 */
export function addActiveStyle( component ) {
	return styled( component )`
		&:active {
			background: ${ colors.$color_button };
			border-color: ${ colors.$color_button_border_hover };
			box-shadow: inset 0 2px 5px -3px ${ rgba( colors.$color_button_border_active, 0.5 ) };
		}
	`;
}

/**
 * Returns a component with applied font size style.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied font size styles.
 */
export function addFontSizeStyles( component ) {
	return styled( component )`
		font-size: 0.8rem;
	`;
}

/**
 * Returns a component with all button selector styles applied.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied styles.
 */
export const addButtonStyles = _flow( [ addBaseStyle, addFocusStyle, addHoverStyle, addActiveStyle ] );

/**
 * Returns a basic styled button.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const BaseButton = addButtonStyles(
	styled.button`
		color: ${ props => props.textColor };
		border-color: ${ props => props.borderColor };
		background: ${ props => props.backgroundColor };
		box-shadow: 0 1px 0 ${ props => rgba( props.boxShadowColor, 1 ) };
	`
);

BaseButton.propTypes = {
	backgroundColor: PropTypes.string,
	textColor: PropTypes.string,
	borderColor: PropTypes.string,
	boxShadowColor: PropTypes.string,
};

BaseButton.defaultProps = {
	backgroundColor: colors.$color_button,
	textColor: colors.$color_button_text,
	borderColor: colors.$color_button_border,
	boxShadowColor: colors.$color_button_border,
	type: "button",
};

/**
 * Returns a styled Button with set font size.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const Button = addFontSizeStyles( BaseButton );

/**
 * Applies styles to icon for IconButton with text.
 *
 * @param {ReactElement} icon The original icon.
 *
 * @returns {ReactElement} Icon with text styles applied.
 */
function addIconTextStyle( icon ) {
	return styled( icon )`
		margin: 0 8px 0 0;
		flex-shrink: 0;
	`;
}

/**
 * Returns an icon button that can optionally contain text.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled icon button.
 */
export const IconButton = ( props ) => {
	const { children: text, icon, iconColor } = props;

	let IconComponent = Icon;
	if ( text ) {
		IconComponent = addIconTextStyle( IconComponent );
	}

	const newProps = omit( props, "icon" );

	return (
		<Button { ...newProps }>
			<IconComponent icon={ icon } color={ iconColor } />
			{ text }
		</Button>
	);
};

IconButton.propTypes = {
	icon: PropTypes.func.isRequired,
	iconColor: PropTypes.string.isRequired,
	children: PropTypes.string,
};
