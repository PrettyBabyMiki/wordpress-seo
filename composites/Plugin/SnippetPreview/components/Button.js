import styled from "styled-components";
import _flow from "lodash/flow";
import colors from "../../../../style-guide/colors.json";
import { rgba } from "../../../../style-guide/helpers";

/**
 * Returns a component with applied focus styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied focus styles.
 */
export function addFocusStyle( component ) {
	return styled( component )`
		&::-moz-focus-inner {
			border-width: 0;
		}

		&:focus {
			outline: none;
			border-color: ${ colors.$color_blue };
			background-color: ${ colors.$color_white };
			box-shadow: 0 0 3px ${ rgba( colors.$color_blue_dark, .8 ) };
		}
	`;
}

/**
 * Returns a component with applied hover styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied hover styles.
 */
export function addHoverStyle( component ) {
	return styled( component )`
		&:hover {
			background: ${ colors.$color_button_hover };
			border-color: ${ colors.$color_button_border_hover };
			color: ${ colors.$color_button_text_hover };
		}
	`;
}

/**
 * Returns a component with applied active styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied active styles.
 */
export function addActiveStyle( component ) {
	return styled( component )`
		&:active {
			background: ${ colors.$color_button };
			border-color: ${ colors.$color_button_border };
			box-shadow: inset 0 2px 5px -3px ${ rgba( colors.$color_button_border_active, 0.5 ) };
		}
	`;
}

/**
 * Returns a component with all button selector styles applied.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied styles.
 */
export const addButtonStyles = _flow( [ addFocusStyle, addHoverStyle, addActiveStyle ] );

/**
 * Returns a basic styled button.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const BaseButton = addButtonStyles(
	styled.button`
		color: ${ colors.$color_button_text };
		border: 1px solid ${ colors.$color_button_border };
		background: ${ colors.$color_button };
		box-shadow: 0 1px 0 ${ rgba( colors.$color_button_border, 1 ) };
		vertical-align: top;
		display: inline-block;
		text-decoration: none;
		padding: 6px 10px 5px;
		border-radius: 3px;
		cursor: pointer;
		box-sizing: border-box;
		font-size: inherit;
		font-family: inherit;
		font-weight: inherit;
	`
);

BaseButton.defaultProps = {
	type: "button",
};

/**
 * Returns a button styled for the SnippetPreview.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const SnippetPreviewButton = styled( BaseButton )`
	line-height: 15px;
	font-size: 0.8rem;
`;
