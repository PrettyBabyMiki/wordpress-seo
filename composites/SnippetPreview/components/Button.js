import styled from "styled-components";

/**
 * @summary:
 * Returns a basic styled button
 *
 * @param {object} props Component props
 *
 * @returns {ReactElement} Styled button
 */
const Button = styled.button`
	display: inline-block;
	padding: 8px 10px;
    border: 1px solid #dbdbdb;
    border-radius: 4px;
    color: #555;
    background: #f7f7f7;
    font-size: 0.8rem;
    line-height: 15px;
    cursor: pointer;
    box-sizing: border-box;
    font-family: inherit;
    font-weight: inherit;
    
    :focus {
		border-color: #5b9dd9;
		outline: none;
		color: #23282d;
		background-color: #fafafa;
		box-shadow: 0 0 3px rgba(0, 115, 170, 0.8);
    }
`;

export default Button;
