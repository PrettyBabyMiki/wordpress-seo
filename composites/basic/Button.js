import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const StyledButton = styled.button`
	display: inline-block;
	padding: 8px 10px;
    border: 1px solid #dbdbdb;
    border-radius: 4px;
    color: #555;
    background: #f7f7f7;
    font-size: 0.8rem;
    line-height: 15px;
    cursor: pointer;
    box-sizing: border-box;
    font-family: inherit;
    font-weight: inherit;
    
    :focus {
		border-color: #5b9dd9;
		outline: none;
		color: #23282d;
		background-color: #fafafa;
		box-shadow: 0 0 3px rgba(0, 115, 170, 0.8);
    }
`;

/**
 * @summary:
 * Returns a basic styled button
 *
 * @param {object} props Component props
 *
 * @returns {ReactElement} Styled button
 */
function Button( props ) {
	return (
		<StyledButton onClick={props.onClick}>
			{props.children}
		</StyledButton>
	);
}

Button.propTypes = {
	onClick: PropTypes.func.isRequired,
	children: PropTypes.any,
};

export default Button;
