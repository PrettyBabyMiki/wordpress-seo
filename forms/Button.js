import React from "react";
import PropTypes from "prop-types";

/**
 * Represents the button HTML element.
 *
 * @param {Object} props The properties to use.
 * @returns {JSX} A representation of the button HTML element based on the passed props.
 * @constructor
 */
const Button = ( props ) => {
	return (
		<button className={ props.className } type="button" onClick={ props.onClick }  { ...props.optionalAttributes }>{ props.text }</button>
	);
};

/**
 * Adds validation for the properties.
 *
 * @type {{text: string, className: string, onClick: function, optionalAttributes: object}}
 */
Button.propTypes = {
	text: PropTypes.string.isRequired,
	className: PropTypes.string,
	onClick: PropTypes.func.isRequired,

	optionalAttributes: PropTypes.object,
};

export default Button;
