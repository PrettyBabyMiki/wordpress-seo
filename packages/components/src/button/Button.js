import React from "react";
import PropTypes from "prop-types";
import "./buttons.css";

export const sharedButtonPropTypes = {
	isLink: PropTypes.bool,
	children: PropTypes.oneOfType(
		[
			PropTypes.node,
			PropTypes.arrayOf( PropTypes.node ),
		]
	),
};

export const sharedButtonDefaultProps = {
	isLink: false,
	children: null,
};

/**
 * Creates a button component.
 *
 * @param {Object} props The props
 *
 * @returns {ReactElemen} The Button component.
 */
const Button = ( props ) => {
	if ( props.isLink ) {
		return <a
			href={ props.href }
			className={ props.className }
		>
			{ props.children }
		</a>;
	}


	return <button
		onClick={ props.onClick }
		className={ props.className }
	>
		{ props.children }
	</button>;
};

Button.propTypes = {
	...sharedButtonPropTypes,
	className: PropTypes.string,
};

Button.defaultProps = {
	...sharedButtonDefaultProps,
	className: "yoast-button",
};

export default Button;
