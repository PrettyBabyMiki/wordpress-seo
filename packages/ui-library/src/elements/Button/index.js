/* eslint-disable no-undefined */
import classNames from "classnames";
import PropTypes from "prop-types";

import Spinner from "../Spinner";

const classNameMap = {
	variant: {
		primary: "yst-button--primary",
		secondary: "yst-button--secondary",
		error: "yst-button--error",
	},
	size: {
		"default": "",
		small: "yst-button--small",
		large: "yst-button--large",
	},
};

/**
 * @param {Object} props Props object.
 * @returns {JSX.Element} Button component.
 */
const Button = ( {
	children,
	as: Component,
	type,
	variant,
	size,
	isLoading,
	isDisabled,
	className,
	...props
} ) => (
	<Component
		type={ type || ( Component === "button" && "button" ) }
		disabled={ isDisabled }
		className={ classNames(
			"yst-button",
			classNameMap.variant[ variant ],
			classNameMap.size[ size ],
			className,
		) }
		{ ...props }
	>
		{ isLoading && <Spinner size={ size === "small" ? 3 : 4 } className="yst-mr-2" /> }
		{ children }
	</Component>
);

Button.propTypes = {
	children: PropTypes.node.isRequired,
	as: PropTypes.elementType,
	type: PropTypes.oneOf( [ undefined, "button", "submit" ] ),
	variant: PropTypes.oneOf( Object.keys( classNameMap.variant ) ),
	size: PropTypes.oneOf( Object.keys( classNameMap.size ) ),
	isLoading: PropTypes.bool,
	isDisabled: PropTypes.bool,
	className: PropTypes.string,
};

Button.defaultProps = {
	as: "button",
	type: undefined,
	variant: "primary",
	size: "default",
	isLoading: false,
	isDisabled: false,
	className: "",
};

export default Button;
