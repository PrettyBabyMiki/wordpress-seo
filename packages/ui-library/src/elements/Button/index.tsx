import React, { ElementType, FC } from  "react";
import cx from "classnames";

import Spinner from "../Spinner";

const classNameMap = {
	variant: {
		primary: "yst-button-primary",
		secondary: "yst-button-secondary",
		error: "yst-button-error",
	},
	size: {
		default: "",
		small: "yst-button-small",
		large: "yst-button-large",
	},
};

interface ButtonProps {
    as?: ElementType;
	type?: "button" | "submit";
	variant?: keyof typeof classNameMap.variant;
	size?: keyof typeof classNameMap.size;
	isLoading?: boolean;
	isDisabled?: boolean;
	className?: string;
	onClick?: () => void;
    [k: string]: unknown; // Wilcard: all props allowed.
}

const Button: FC<ButtonProps> = ( {
	children,
	as: Component = "button",
	type = "button",
	variant = "primary",
	size = "default",
	isLoading = false,
	isDisabled = false,
	className = "",
	...props
} ) => (
	<Component
		type={ type }
		disabled={ isDisabled }
		className={ cx(
			"yst-button",
			classNameMap.variant[ variant ],
			classNameMap.size[ size ],
			className,
		) }
		{ ...props }
	>
		{ isLoading && <Spinner size={ size === "small" ? 3 : 4 } className="yst-mr-2" /> }
		{ children }
	</Component>
);

export default Button;