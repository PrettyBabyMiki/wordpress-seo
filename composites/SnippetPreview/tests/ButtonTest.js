import React from "react";
import renderer from "react-test-renderer";

import Button from "../components/Button";

test( "the button matches the snapshot", () => {
	const component = renderer.create(
		<Button>ButtonValue</Button>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "button executes callback", () => {
	const component = renderer.create(
		<Button onClick={
			() => {
				return "clicked";
			}
		}>ButtonValue</Button>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();

	tree.props.onClick();

	tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
