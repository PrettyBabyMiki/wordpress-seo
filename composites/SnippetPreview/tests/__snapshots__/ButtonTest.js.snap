// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button executes callback 1`] = `
<button
  className="sc-bxivhb jdkbtw sc-htpNat clPCcB sc-bwzfXH kGsEoO sc-bdVaJa hZtEcp"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`Button executes callback 2`] = `
<button
  className="sc-bxivhb jdkbtw sc-htpNat clPCcB sc-bwzfXH kGsEoO sc-bdVaJa hZtEcp"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`SnippetPreviewButton executes callback 1`] = `
<button
  className="sc-ifAKCX hpLiYA sc-bxivhb jdkbtw sc-htpNat clPCcB sc-bwzfXH kGsEoO sc-bdVaJa hZtEcp"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`SnippetPreviewButton executes callback 2`] = `
<button
  className="sc-ifAKCX hpLiYA sc-bxivhb jdkbtw sc-htpNat clPCcB sc-bwzfXH kGsEoO sc-bdVaJa hZtEcp"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`the Button matches the snapshot 1`] = `
<button
  className="sc-bxivhb jdkbtw sc-htpNat clPCcB sc-bwzfXH kGsEoO sc-bdVaJa hZtEcp"
  type="button"
>
  ButtonValue
</button>
`;

exports[`the SnippetPreviewButton matches the snapshot 1`] = `
<button
  className="sc-ifAKCX hpLiYA sc-bxivhb jdkbtw sc-htpNat clPCcB sc-bwzfXH kGsEoO sc-bdVaJa hZtEcp"
  type="button"
>
  ButtonValue
</button>
`;
